### ISSUE TITLE
v4: version and clock sequence bits not properly set

### ISSUE THREAD

**Issue by AaronM04** [as/uuid#5]
The high nibble of this byte should be 4 for it to be a valid V4 UUID: https://github.com/as/uuid/blob/f2107c8d81ca77e383bbf16fd25394a2bbddf303/uuid.go#L78

**Comment by AaronM04** [comment]
It appears that replacing `h[c[6]&15]` with `'4'` would fix the issue.

**Comment by as** [comment]
It's even more elaborate than that, unfortunately. The RFC also says to set bits 6 and 7 to zero and one. The question is whether anyone practically checks for these two things and whether other generators follow the RFC correctly.

https://www.ietf.org/rfc/rfc4122.txt

```
4.4.  Algorithms for Creating a UUID from Truly Random or
      Pseudo-Random Numbers

   The version 4 UUID is meant for generating UUIDs from truly-random or
   pseudo-random numbers.

   The algorithm is as follows:

   o  Set the two most significant bits (bits 6 and 7) of the
      clock_seq_hi_and_reserved to zero and one, respectively.

   o  Set the four most significant bits (bits 12 through 15) of the
      time_hi_and_version field to the 4-bit version number from
      Section 4.1.3.

   o  Set all the other bits to randomly (or pseudo-randomly) chosen
      values.
```

**Comment by AaronM04** [comment]
So an additional step would be needed to fix it: change the next line (L79) to read:

```go
h[c[7]&15], h[((c[7]>>4)&3) + 8],
```

**Comment by as** [comment]
```
   0                   1                   2                   3
    0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |                          time_low                             |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |       time_mid                |         time_hi_and_version   |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |clk_seq_hi_res |  clk_seq_low  |         node (0-1)            |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |                         node (2-5)                            |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

It looks like it would actually be on `c[8]`. These are generated as ASCII output, so it would probably be simpler to create a second table and have:

`var h8 = [...]byte{'8', '8', '8', '8', '9', '9', '9', '9', 'a', 'a', 'a', 'a', 'b', 'b', 'b', 'b'}`

The binary nibble must start with `10`, so it constrains us to the base 16 set `{8, 9, a, b}`

**Comment by as** [comment]
We can ignore the part about the second table, that's overly complex. I used bit-wise operations to pigeonhole the numbers to select from `{8, 9, a, b}`

**Comment by as** [comment]
Fixed by pull #6
